labs/b6.md

```diff
@@ -0,0 +1,153 @@
+29756 pts/2    00:00:00 sleep
+29757 pts/2    00:00:00 ps
+```
+In the **first** terminal run `ps` again. You should notice that the `sleep` process is not showing up, even though the thousand seconds haven't expired. Why do you think this behavior occurs (hint: TTY column)?
+
+We can get the process to display on the first terminal by running `ps -u`, which displays all the processes running as your user. Notice the PID column; each process has a unique ID assigned to it by the kernel. One thing we can do with this PID is send signals to the process. `sleep 1000` is pretty useless, so go ahead and kill it -- `kill 29756` (substitute `29756`  with whatever `ps` outputted for you).
+
+The most common use of `ps` is to run `ps -ef` to see all the processes running on the system. Run `ps -e` and `ps -f` independently to see how the flags work together.
+
+## htop
+Open up a terminal and run the `htop` command. `htop` can be thought of as a more extensive version of `ps -ef`, whereby process stats are updated in real-time.
+
+First press `<F2>`, scroll down to Display options, and check "Hide userland process threads." We won’t be dealing with those in this lab.
+
+Now open up another terminal and SSH into your VM. Run the command `yes`. It uses a lot of resources as it prints a continuous stream of `y`'s. What resource specifically does the `yes` command exhaust? If you are having trouble finding this, press `<` to choose which resource to order processes by. Make sure to quit out of `yes` (^C) once you are finished.
+
+### The process hierarchy
+Run `htop` once more. This time click `<F5>` to enter Tree View. You should see a visual representation of the process hierarchy on your system, with everything stemming from `/sbin/init` (systemd).
+
+For curious students that are interested in seeing a more extensive process hierarchy on a large system, you are encouraged to run `htop` on the OCF server `tsunami`. Let us know of any cool processes that you find!
+
+### Orphan processes
+Open a second terminal and SSH to your VM. Now run `sleep 1000 &`. You should see this new process pop into your `htop` session on your first terminal. If not, press `<F3>` and search for "sleep." What is its parent?
+
+Select this parent and press `<F9>` to kill it. Send the `SIGTERM` signal. The sleep process now has `init` as its new parent, which is PID 1. What you just did is manually orphan a process; when that happens said process is subsequently re-parented by the `init` process.
+
+Now go through the same steps again. This time, send the parent a `SIGHUP` (hangup) signal. Can you still find the sleep process? When `SIGHUP` is sent to a parent shell, the parent subsequently sends hangup signals to any child processes before terminating; all processes that receive `SIGHUP` from a parent shell will terminate -- this is one way to avoid creating orphan processes.
+
+If you are interested in learning about the different signals, run `man 7 signal`.
```

Comment:
may also want to explain where the 7 in this comes from

Reply:
This is a processes and services lab. `man man` should have been covered in in the first couple lectures.

Reply:
It doesn't, actually. Just linking to like https://stackoverflow.com/questions/62936/what-does-the-number-in-parentheses-shown-after-unix-command-names-in-manpages-m would probably be helpful

Reply:
Looks like the exact same thing as running `man man`, which I'll include. Though I'll also add this as a todo for our intro to shell lecture as the manpage that tells one how to navigate man pages is pretty important.